crates/provider/src/native.rs

```diff
@@ -1,833 +1,1522 @@
-            return Err(ProviderError::RunCommandError(args.join(" ")));
-        } else {
-            // cmd success or we allow to fail
-            // in either case we return Ok
-            Ok(RunCommandResponse {
-                exit_code: result.status,
-                std_out: String::from_utf8_lossy(&result.stdout).into(),
-                std_err: if result.stderr.is_empty() {
-                    None
-                } else {
-                    Some(String::from_utf8_lossy(&result.stderr).into())
-                },
+        // store node inside namespace
+        inner.nodes.insert(options.name, node.clone());
+
+        Ok(Arc::new(node))
+    }
+
+    async fn generate_files(&self, options: GenerateFilesOptions) -> Result<(), ProviderError> {
+        // we spawn a node doing nothing but looping so we can execute our commands
+        let temp_node = self
+            .spawn_node(SpawnNodeOptions {
+                name: format!("temp_{}", Uuid::new_v4()),
+                command: "bash".to_string(),
+                args: vec!["-c".to_string(), "while :; do sleep 1; done".to_string()],
+                env: vec![],
+                injected_files: options.injected_files,
             })
+            .await?;
+
+        for GenerateFileCommand {
+            command,
+            args,
+            env,
+            local_output_path,
+        } in options.commands
+        {
+            match temp_node
+                .run_command(RunCommandOptions { command, args, env })
+                .await
+                .map_err(|err| ProviderError::FileGenerationFailed(err.into()))?
+            {
+                Ok(contents) => self
+                    .filesystem
+                    .write(
+                        format!(
+                            "{}{}",
+                            self.base_dir.to_string_lossy(),
+                            local_output_path.to_string_lossy()
+                        ),
+                        contents,
+                    )
+                    .await
+                    .map_err(|err| ProviderError::FileGenerationFailed(err.into()))?,
+                Err((_, msg)) => Err(ProviderError::FileGenerationFailed(anyhow!("{msg}")))?,
+            };
         }
+
+        temp_node.destroy().await
     }
 
-    // TODO: Add test
-    async fn run_script(
-        &mut self,
-        identifier: String,
-        script_path: String,
-        args: Vec<String>,
-    ) -> Result<RunCommandResponse, ProviderError> {
-        let script_filename = Path::new(&script_path)
-            .file_name()
-            .ok_or(ProviderError::InvalidScriptPath(script_path.clone()))?
-            .to_str()
-            .ok_or(ProviderError::InvalidScriptPath(script_path.clone()))?;
-        let script_path_in_pod = format!("{}/{}/{}", self.tmp_dir, identifier, script_filename);
+    async fn static_setup(&self) -> Result<(), ProviderError> {
```

Comment:
Add a TODO Note maybe on what needs to be done?

Reply:
I added a small comment :+1: